### ISSUE TITLE
Remove NestedTensor

### ISSUE THREAD

**Issue by whatdhack** [facebookresearch/detr#233]
## Instructions To Reproduce the 🐛 Bug:

Is there a way to replace the NestedTensor with 2 tensors - one for images and other for masks , without having to retrain the network ?  The use of this odd class as input cuases lot of headache . 

## Expected behavior:

replace NestedTensor with 2 tensors. 

## Environment:
1.6

**Comment by zhiqwang** [comment]
I think the question of the first importance is what the scenario you are facing. If you are processing the object detection problem, you can just drop the `NestedTensor` away, which corresponds the code in

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/models/detr.py#L59-L60

and 

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/models/backbone.py#L72-L73

Because the backbone just use the `Tensor` information of the input. In this conditions, it's easy. Otherwise in the segmentation, it seems that we should do a lot to detach the object `NestedTensor` from the architecture (In my opinion, the `NestedTensor` here was just designed to handling the complicated segmentation tasks).

**Comment by whatdhack** [comment]
Sounds like a good  idea, however on further  look,  "mask" is used later on in "out", which in turn is used  later. 

```
    def forward(self, tensor_list: NestedTensor):
        xs = self.body(tensor_list.tensors)
        out: Dict[str, NestedTensor] = {}
        for name, x in xs.items():
            m = tensor_list.mask
            assert m is not None
            mask = F.interpolate(m[None].float(), size=x.shape[-2:]).to(torch.bool)[0]
            out[name] = NestedTensor(x, mask)
        return out
```

```
        if isinstance(samples, (list, torch.Tensor)):
            samples = nested_tensor_from_tensor_list(samples)
        features, pos = self.backbone(samples)

        src, mask = features[-1].decompose()
        assert mask is not None
        hs = self.transformer(self.input_proj(src), mask, self.query_embed.weight, pos[-1])[0]
```

**Comment by whatdhack** [comment]
This code looks more like C++ than python - are all these typing really ncecessary ?

**Comment by zhiqwang** [comment]
> This code looks more like C++ than python - are all these typing really ncecessary ?

Some of these typing is for converting to torchscripts, it's necessary in this situation.

**Comment by zhiqwang** [comment]
> ..., however on further look, "mask" is used later on in "out", which in turn is used later.

You are right. I have confused the usage of `NestedTensor` with my own repo :( . In my own understanding, DETR packages the `NestedTensor` using the function `nested_tensor_from_tensor_list` as below.

https://github.com/facebookresearch/detr/blob/5e66b4cd15b2b182da347103dd16578d28b49d69/util/misc.py#L283-L305

If you want to drop `NestedTensor` out, it means that you should reimplement this function is another ways.

**Comment by fmassa** [comment]
It is possible to remove NestedTensor without having to retrain the model, but will require a couple of changes to the code (which shouldn't be too difficult and most has already been pointed out).

For more background on NestedTensor, see https://github.com/facebookresearch/detr/issues/116

I believe I've answered your question, and as such I'm closing this issue, but let us know if you have further questions

**Comment by whatdhack** [comment]
Who knows when the Python NestedTensor going to finalized, whether every vendor decides to support that  It is probably years away,    Would request figuring out a better way than just relying on uncertain future work.
